libcpuid/rdmsr.c

```diff
@@ -31,67 +31,99 @@
 #include "libcpuid_util.h"
 #include "rdtsc.h"
 
-#ifndef _WIN32
-#  ifdef __APPLE__
-/* On Darwin, we still do not support RDMSR, so supply dummy struct
-   and functions */
-struct msr_driver_t { int dummy; };
+#if defined (__linux__) || defined (__gnu_linux__)
+/* Assuming linux with /dev/cpu/x/msr: */
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+struct msr_driver_t { int fd; };
+static int rdmsr_supported(void);
 struct msr_driver_t* cpu_msr_driver_open(void)
 {
-	set_error(ERR_NOT_IMP);
-	return NULL;
+	return cpu_msr_driver_open_core(0);
 }
 
-struct msr_driver_t* cpu_msr_driver_open_core(int core_num)
+struct msr_driver_t* cpu_msr_driver_open_core(uint8_t core_num)
 {
-	set_error(ERR_NOT_IMP);
-	return NULL;
+	char msr[32];
+	struct msr_driver_t* handle;
+	if (core_num >= cpuid_get_total_cpus())	{
+		set_error(ERR_INVCNB);
+		return NULL;
+	}
+	if (!rdmsr_supported()) {
+		set_error(ERR_NO_RDMSR);
+		return NULL;
+	}
+	sprintf(msr, "/dev/cpu/%i/msr", core_num);
+	int fd = open(msr, O_RDONLY);
+	if (fd < 0) {
+		if (errno == EIO) {
+			set_error(ERR_NO_RDMSR);
+			return NULL;
+		}
+		set_error(ERR_NO_DRIVER);
+		return NULL;
+	}
+	handle = (struct msr_driver_t*) malloc(sizeof(struct msr_driver_t));
+	handle->fd = fd;
+	return handle;
 }
 
-int cpu_rdmsr(struct msr_driver_t* driver, int msr_index, uint64_t* result)
+int cpu_rdmsr(struct msr_driver_t* driver, uint32_t msr_index, uint64_t* result)
 {
-	return set_error(ERR_NOT_IMP);
-}
+	ssize_t ret;
 
-int cpu_msr_driver_close(struct msr_driver_t* driver)
-{
-	return set_error(ERR_NOT_IMP);
+	if (!driver || driver->fd < 0)
+		return set_error(ERR_HANDLE);
+	ret = pread(driver->fd, result, 8, msr_index);
+	if (ret != 8)
+		return set_error(ERR_INVMSR);
+	return 0;
 }
 
-#define MSRINFO_DEFINED
-int cpu_msrinfo(struct msr_driver_t* driver, cpu_msrinfo_request_t which)
+int cpu_msr_driver_close(struct msr_driver_t* drv)
 {
-	return set_error(ERR_NOT_IMP);
+	if (drv && drv->fd >= 0) {
+		close(drv->fd);
+		free(drv);
+	}
+	return 0;
 }
-#  else /* __APPLE__ */
-/* Assuming linux with /dev/cpu/x/msr: */
-#include <unistd.h>
-#include <stdio.h>
-#include <sys/types.h>
-#include <sys/stat.h>
+
+/* #endif defined (__linux__) || defined (__gnu_linux__) */
+
+#elif defined (__FreeBSD__) || defined (__DragonFly__)
+/* Assuming FreeBSD with /dev/cpuctlX */
 #include <fcntl.h>
+#include <unistd.h>
 #include <errno.h>
+#include <sys/ioctl.h>
+#include <sys/cpuctl.h>
+
 struct msr_driver_t { int fd; };
 static int rdmsr_supported(void);
 struct msr_driver_t* cpu_msr_driver_open(void)
 {
 	return cpu_msr_driver_open_core(0);
 }
 
-struct msr_driver_t* cpu_msr_driver_open_core(int core_num)
+struct msr_driver_t* cpu_msr_driver_open_core(uint8_t core_num)
```

Comment:
Cool, perfect. I'd write up some docs on the MSR usage requirements in terms of user permissions, etc., to have that stated officially.